File: data-structures/lists/python/nested_lists_find_astronaut_name.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0


# Index values of each astronaut
CRAFT_INDEX = 0
NAME_INDEX = 1
DAYS_INDEX = 2


def find_position_of_astronaut(all_astronauts, search_name):
    """Return the index of the astronaut that contains the given name"""

    i = 0
    found_index = -1  # Return -1 if no astronaut found

    # Repeat while the astronaut has not been found
    while found_index == -1 and i < len(all_astronauts):
        # Check if the name value contains the search name substring
        if search_name in all_astronauts[i][NAME_INDEX]:
            # Stop searching and return the index
            found_index = i
        i += 1 # Increment the counter
       
    return found_index


def display_astronaut_details(astronaut):
    """Display the values from the given astronaut list"""

    print("\nAstronaut details:")
    
    # Repeat for each value in the list
    for value in astronaut:
        print(value)


def main():
    # Declare and initialise the nested list of astronauts
    all_astronauts = [
        ["ISS", "Sergey Prokopyev", "176"],
        ["ISS", "Dmitry Petelin", "176"],
        ["ISS", "Frank Rubio", "176"],
        ["Tiangong", "Fei Junlong", "112"],
        ["Tiangong", "Deng Qingming", "107"],
        ["Tiangong", "Zhang Lu", "107"],
        ["ISS", "Stephen Bowen", "54"],
        ["ISS", "Warren Hoburg", "14"],
        ["ISS", "Sultan Al Neyadi", "14"],
        ["ISS", "Andrey Fedyaev", "14"]
    ]
    
    # Check if one of the astronaut lists contains a name substring
    name_to_find = input("Enter the name of an astronaut to find: ")
    position = find_position_of_astronaut(all_astronauts, name_to_find)
    
    # Output the search results
    if position == -1:
        print("Astronaut not found in this data set")
    else:
        print(f"Astronaut found at position: {position}")
        display_astronaut_details(all_astronauts[position])


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()

File: data-structures/lists/java/nested_lists_find_astronaut_name.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.ArrayList;
import java.util.Scanner;

class NestedLists  {
    // Index values of each astronaut
    public static final int CRAFT_INDEX = 0;
    public static final int NAME_INDEX = 1;
    public static final int DAYS_INDEX = 2;

    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        // Declare and initialise the nested list of astronauts
        ArrayList<ArrayList<String>> allAstronauts = new ArrayList<ArrayList<String>>();
        initialiseAstronauts(allAstronauts);

        // Use Scanner for reading user input from the console
        Scanner scanner = new Scanner(System.in);

        // Check if one of the astronaut lists contains a name substring
        System.out.println("Enter the name of an astronaut to find:");
        String nameToFind = scanner.nextLine();
        int position = findPositionOfAstronaut(allAstronauts, nameToFind);

        // Output the search results
        if (position == -1) {
            System.out.println("Astronaut not found in this data set");
        }
        else {
            System.out.println("Astronaut found at position " + position);
            displayAstronautDetails(allAstronauts.get(position));
        }
    }

    // Return the index of the astronaut that contains the search name
    public static int findPositionOfAstronaut(ArrayList<ArrayList<String>> allAstronauts,
                                              String searchName) {

        int i = 0;
        int foundIndex = -1;  // Return -1 if no astronaut found

        // Repeat while the astronaut has not been found
        while (foundIndex == -1 && i < allAstronauts.size()) {
            // Check if the name value contains the search name substring
            if (allAstronauts.get(i).get(NAME_INDEX).contains(searchName)) {
                // Stop searching and return the index
                foundIndex = i;
            }
            i += 1; // Increment the counter
        }

        return foundIndex;
    }

    // Display the values from the given astronaut list
    public static void displayAstronautDetails(ArrayList<String> astronaut) {
        System.out.println("\nAstronaut details:");

        // Repeat for each value in the list
        for (String value : astronaut) {
            System.out.println(value);
        }
    }

    // Initialise the nested list of astronauts
    public static void initialiseAstronauts(ArrayList<ArrayList<String>> allAstronauts) {
        ArrayList<String> first = new ArrayList<String>();
        first.add("ISS");
        first.add("Sergey Prokopyev");
        first.add("176");
        allAstronauts.add(first);

        ArrayList<String> second = new ArrayList<String>();
        second.add("ISS");
        second.add("Dmitry Petelin");
        second.add("176");
        allAstronauts.add(second);

        ArrayList<String> third = new ArrayList<String>();
        third.add("ISS");
        third.add("Frank Rubio");
        third.add("176");
        allAstronauts.add(third);

        ArrayList<String> fourth = new ArrayList<String>();
        fourth.add("Tiangong");
        fourth.add("Fei Junlong");
        fourth.add("112");
        allAstronauts.add(fourth);

        ArrayList<String> fifth = new ArrayList<String>();
        fifth.add("Tiangong");
        fifth.add("Deng Qingming");
        fifth.add("107");
        allAstronauts.add(fifth);

        ArrayList<String> sixth = new ArrayList<String>();
        sixth.add("Tiangong");
        sixth.add("Zhang Lu");
        sixth.add("107");
        allAstronauts.add(sixth);

        ArrayList<String> seventh = new ArrayList<String>();
        seventh.add("ISS");
        seventh.add("Stephen Bowen");
        seventh.add("54");
        allAstronauts.add(seventh);

        ArrayList<String> eighth = new ArrayList<String>();
        eighth.add("ISS");
        eighth.add("Warren Hoburg");
        eighth.add("14");
        allAstronauts.add(eighth);

        ArrayList<String> ninth = new ArrayList<String>();
        ninth.add("ISS");
        ninth.add("Sultan Al Neyadi");
        ninth.add("14");
        allAstronauts.add(ninth);

        ArrayList<String> tenth = new ArrayList<String>();
        tenth.add("ISS");
        tenth.add("Andrey Fedyaev");
        tenth.add("14");
        allAstronauts.add(tenth);
    }
}